pkg/collector/tcp.go

```diff
@@ -63,48 +64,36 @@ func (cp *CollectingProcess) handleTCPClient(conn net.Conn) {
 	address := conn.RemoteAddr().String()
 	client := cp.createClient()
 	cp.addClient(address, client)
-	buff := make([]byte, cp.maxBufferSize)
 	defer cp.wg.Done()
 	defer conn.Close()
 	go func() {
+		// maxBufferSize should be good enough to hold maximum sized IPFIX packet.
+		buff := make([]byte, cp.maxBufferSize)
+		reader := bufio.NewReader(conn)
 		for {
-			size, err := conn.Read(buff)
+			length, err := getMessageLength(reader)
 			if err != nil {
-				if err == io.EOF {
-					klog.Infof("Connection from %s has been closed.", address)
-				} else {
-					klog.Errorf("Error in collecting process: %v", err)
-				}
+				klog.Errorf("error when retreiving message length: %v", err)
 				cp.deleteClient(address)
 				return
 			}
-			klog.V(2).Infof("Receiving %d bytes from %s", size, address)
-			buffBytes := make([]byte, size)
-			copy(buffBytes, buff[:size])
-			for size > 0 {
-				length, err := getMessageLength(bytes.NewBuffer(buffBytes))
-				if err != nil {
-					klog.Error(err)
-					buffBytes = nil
-					break
-				}
-				if size < length {
-					klog.Errorf("Message length %v is larger than size read from buffer %v", length, size)
-					buffBytes = nil
-					break
-				}
-				size = size - length
-				// get the message here
-				message, err := cp.decodePacket(bytes.NewBuffer(buffBytes[0:length]), address)
-				if err != nil {
-					klog.Error(err)
-					buffBytes = nil
-					break
-				}
-				klog.V(4).Infof("Processed message from exporter %v, number of records: %v, observation domain ID: %v",
-					message.GetExportAddress(), message.GetSet().GetNumberOfRecords(), message.GetObsDomainID())
-				buffBytes = buffBytes[length:]
+			_, err = io.ReadFull(reader, buff[:length])
```

Comment:
you don't need the full-sized `buff` at all, and the copy is not needed.
this should read:
```go
buff := make([]byte, length)
_, err = io.ReadFull(reader, buff)
```

at that stage, and based on the rest of the code, I don't think you will have any superfluous copies. It is a bit of  a shame that we still have to do a dynamic buffer allocation for each packet, but there is no _easy_ way around that. The next logical step would be to use a large ring buffer and avoid these dynamic allocations.

Reply:
Agreed.. we can avoid the copy. This means that there is no buffer required with `MaxBufferSize` which is specified by the user. I have seen some improvements, especially with CPU/runtime because of no copy.

Thanks for the insight on dynamic buffer allocation. We will look into ring bugger approach in a follow up PR.